distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+        # id(ws) is there just to prevent WorkerState objects to be compared in the
+        # unlikely event that they have exactly the same amount of bytes allocated.
+        _, _, ws = max((self.workers_memory[ws], id(ws), ws) for ws in candidates)
+        return ws
+
+
+class ActiveMemoryManagerPlugin:
+    """Abstract class"""
+
+    manager: ActiveMemoryManager
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}()"
+
+    def run(
+        self,
+    ) -> """Generator[
+        tuple[str, TaskState, Optional[set[WorkerState]]],
+        Optional[WorkerState],
+        None
+    ]""":
+        """This method is invoked by the ActiveMemoryManager every few seconds.
+        It is an iterator that must emit one of the following:
+
+        - "replicate", <TaskState>, None
+        - "replicate", <TaskState>, {subset of potential workers to replicate to}
+        - "drop", <TaskState>, None
+        - "drop", <TaskState>, {subset of potential workers to drop from}
+
+        Each element yielded indicates the desire to create or destroy a single replica
+        of a key. Either the ActiveMemoryManager or the Worker may later decide to
+        disregard the request, e.g. because it would delete the last copy of a key or
+        because the key is currently needed on that worker.
+
+        You may optionally acquire the decision of which worker the key will be
+        replicated on or dropped from, as follows:
+
+        ```python
+        choice = (yield "replicate", ts, None)
+        ```
+        ``choice`` is either a WorkerState or None; the latter is if the
+        ActiveMemoryManager chose to disregard the request.
+
+        The current pending (accepted) commands can be inspected on
+        ``self.manager.pending``; this includes the commands previously yielded by this
+        same method.
+
+        The current memory usage on each worker, *downstream of all pending commands*,
+        can be inspected on ``self.manager.workers_memory``.
+        """
+        raise NotImplementedError("Virtual method")
+
+
+class ReplicateReduce(ActiveMemoryManagerPlugin):
+    """Make sure that we don't have more replicas than desired for every key; drop
+    excess replicas
+    """
+
+    def run(self):
+        # TODO this is O(n) to the total number of in-memory tasks on the cluster; it
+        #      could be made faster by automatically attaching it to a key when it goes
+        #      above one replica and detaching it when it goes below.
+        for ts in self.manager.scheduler.tasks.values():
+            desired_replicas = 1  # TODO have a marker on TaskState
+            ndrop = len(ts.who_has) - max(desired_replicas, len(ts.waiters))
+            if ts in self.manager.pending:
+                pending_repl, pending_drop = self.manager.pending[ts]
+                ndrop += len(pending_repl) - len(pending_drop)
+            for _ in range(ndrop):
+                yield "drop", ts, None
+
```

Comment:
everything below this line is out of scope for the final version of this PR; I'm going to delete it before the final review and re-add it progressively over several other PRs